test/recipes/82-test_ech_client_server.t

```diff
@@ -97,22 +133,85 @@ sub start_ech_client_server
+                         "-ech_outer_alpn", "http451",
+                         "-prexit");
+    } elsif ($test_type eq "GREASE-type" ) {
+        # GREASE with suite
+        @s_client_cmd = ("s_client",
+                         "-connect", "localhost:$s_server_port",
+                         "-servername", "server.example",
+                         "-CAfile", $root_pem,
+                         "-ech_grease_type", "12345",
+                         "-prexit");
+    } elsif ($test_type eq "GREASE" ) {
+        # GREASE with suite
+        @s_client_cmd = ("s_client",
+                         "-connect", "localhost:$s_server_port",
+                         "-servername", "server.example",
+                         "-CAfile", $root_pem,
+                         "-ech_grease",
+                         "-prexit");
+    } elsif ($test_type eq "no-outer" ) {
+        # Real ECH, no outer SNI
+        @s_client_cmd = ("s_client",
+                         "-connect", "localhost:$s_server_port",
+                         "-servername", "server.example",
+                         "-CAfile", $root_pem,
+                         "-ech_config_list", $good_b64,
+                         "-ech_no_outer_sni",
+                         "-prexit");
+    } elsif ($test_type eq "bad-ech" ) {
+        # bad ECH
+        @s_client_cmd = ("s_client",
+                         "-connect", "localhost:$s_server_port",
+                         "-servername", "server.example",
+                         "-CAfile", $root_pem,
+                         "-ech_config_list", "AEH+DQA91wAgACCBdNrnZxqNrUXSyimqqnfmNG4lHtVsbmaaIeRoUoFWFQAEAAEAAQAOc2VydmVyLmV4YW1wbGUAAA==",
+                         "-prexit");
+    } elsif ($test_type eq "cid-free" ) {
+        # Real ECH, ignore CID
+        @s_client_cmd = ("s_client",
+                         "-connect", "localhost:$s_server_port",
+                         "-servername", "server.example",
+                         "-CAfile", $root_pem,
+                         "-ech_config_list", $good_b64,
+                         "-ech_ignore_cid",
+                         "-prexit");
+    } elsif ($test_type eq "cid-wrong" ) {
+        # Real ECH, ignore CID, no trial decrypt
         @s_client_cmd = ("s_client",
-                            "-connect", "localhost:$s_server_port",
-                            "-servername", "server.example",
-                            "-CAfile", $root_pem,
-                            "-ech_config_list", $b64ech,
-                            "-prexit");
-                        # for loadsa debugging add...
-                        # "-debug", "-msg", "-trace", "-tlsextdebug");
+                         "-connect", "localhost:$s_server_port",
+                         "-servername", "server.example",
+                         "-CAfile", $root_pem,
+                         "-ech_config_list", $good_b64,
+                         "-ech_ignore_cid",
+                         "-prexit");
```

Comment:
I'm struggling to see the difference between the "cid-free" test and this one. They look the same? Optionally we could just combine them `($test_type eq "cid-free" ) || ($test_type eq "cid-wrong" )`

Trying to decide if this is a good idea or not. So approved either way.

Reply:
As a BTW, I'll be working in the background on bringing over more tests and writing up documentation, that'll be in [this branch](https://github.com/sftcd/openssl/tree/tests-docs) which'll likely be the source for the next feature branch PR. That's not likely to be a PR immediately though, as there's a bit of work to change test code that predates the set of agreed ECH APIs etc.